[ARTEMIS-4159] Support duplicate cache size configuration per address

fartzy opened the pull request:
### **Why were these changes made?**
This pull request introduces support for configuring a specific id-cache-size per address - [ARTEMIS-4159](https://issues.apache.org/jira/browse/ARTEMIS-4159). The changes include modifications to various files:
* artemis-server/src/main/java/org/apache/activemq/artemis/core/deployers/impl/FileConfigurationParser.java
* artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java
* artemis-server/src/main/java/org/apache/activemq/artemis/core/settings/impl/AddressSettings.java
* artemis-server/src/main/resources/schema/artemis-configuration.xsd
* artemis-server/src/test/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImplTest.java
* artemis-server/src/test/java/org/apache/activemq/artemis/core/config/impl/FileConfigurationTest.java
* artemis-server/src/test/java/org/apache/activemq/artemis/core/settings/AddressSettingsTest.java
* artemis-server/src/test/resources/ConfigurationTest-full-config.xml
* artemis-server/src/test/resources/ConfigurationTest-xinclude-config-address-settings.xml
* artemis-server/src/test/resources/ConfigurationTest-xinclude-schema-config-address-settings.xml
* docs/user-manual/en/address-settings.md
* docs/user-manual/en/configuration-index.md
* docs/user-manual/en/duplicate-detection.md
* tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/DuplicateDetectionTest.java

**Changes Made:**
* FileConfigurationParser.java
Added a constant ID_CACHE_SIZE to represent the id-cache-size attribute in the XML configuration.
Updated the FileConfigurationParser to parse the id-cache-size attribute from the XML and set it on the AddressSettings.
* PostOfficeImpl.java
Introduced a new method resolveIdCacheSize to determine the effective id-cache-size for an address.
Updated getDuplicateIDCache method to use the resolveIdCacheSize method to get the proper id-cache-size for the address.
* AddressSettings.java
Added a new attribute idCacheSize to store the id-cache-size configuration for an address.
Implemented getter and setter methods for idCacheSize.
* artemis-configuration.xsd
Updated the XSD schema to include the id-cache-size attribute for the address settings.
* ConfigurationImplTest.java
Added a new test method to verify parsing of id-cache-size configuration from prefixed properties.
* FileConfigurationTest.java
Added a new test method to verify the parsing of the id-cache-size configuration from the XML configuration file.
* ConfigurationTest-full-config.xml, ConfigurationTest-xinclude-config-address-settings.xml, ConfigurationTest-xinclude-schema-config-address-settings.xml
Added <id-cache-size> elements to the address settings to test the configuration.
* DuplicateDetectionTest.java
Added a new test method to verify the functionality of the address-specific id-cache-size setting.
* AddressSettingsTest.java 
Added to the testSingleMerge to make sure idCacheSize merges correctly.

* docs/user-manual/en/address-settings.md, docs/user-manual/en/configuration-index.md, docs/user-manual/en/duplicate-detection.md
Added documentation for the above for address specific information about `id-cache-size`.

**User Facing Changes:**
New Configuration Option: id-cache-size for Address Settings
With this pull request, administrators now have the ability to configure a specific id-cache-size for each address, allowing better customization and optimization of resources for different addresses.


**Motivation:**
This enhancement provides the ability to configure a specific id-cache-size for each address, allowing better customization and optimization of resources for different addresses. Administrators can now specify the cache size for individual addresses, which will default to the global setting if not explicitly set for an address. This improvement eliminates the need for deploying multiple instances of Artemis with different global cache sizes based on address requirements. The pull request aims to simplify configuration and reduce resource consumption in scenarios where different addresses have different caching needs.

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/settings/impl/AddressSettings.java:
```diff
@@ -1797,6 +1817,8 @@ public void encode(ActiveMQBuffer buffer) {
 
       BufferHelper.writeNullableLong(buffer, pageLimitMessages);
 
+      BufferHelper.writeNullableInteger(buffer, idCacheSize);
```
The writing and reading of the nullable `Integer` is not in the same order so decoding will fail.

fartzy replied:
I moved the read up to the 4th ordinal position from the bottom so they are both same now. Thank you for catching.

jbertram replied:
Both the read and the write for the `idCacheSize` should be at the *end* otherwise the broker may choke on existing persisted address settings when upgrading.

fartzy replied:
oh yeah it will definitely get mixed up with currently persisted messages - thanks. I put both the read and write at end now

michaelandrepearce commented:
Nice contribution, i see justins spotted already one issue to be fixed.
I would ask if we can also update documentation for this configuration and feature.

fartzy commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java:
```diff
@@ -1407,9 +1407,16 @@ public Pair<RoutingContext, Message> redistribute(final Message message,
       return null;
    }
 
+   // ARTEMIS-4159
```
I took out this comment

fartzy commented:
> 

I added some doc to three of the `.md` files. Wherever `id-cache-size` is in the doc already then I added an address specific setting to that.  And also I added to the general `address-settings.md`.

jbertram commented:
All commits should be squashed and the commit message should follow the pattern outlined in the [Hacking Guide](https://github.com/apache/activemq-artemis/blob/main/docs/hacking-guide/en/code.md#commitMessageDetails).

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/deployers/impl/FileConfigurationParser.java:
```diff
@@ -1456,6 +1458,10 @@ protected Pair<String, AddressSettings> parseAddressSettings(final Node node) {
             addressSettings.setEnableIngressTimestamp(XMLUtil.parseBoolean(child));
+         } else if (ID_CACHE_SIZE.equalsIgnoreCase(name)) {
+            int idCacheSize = XMLUtil.parseInt(child);
+            Validators.GT_ZERO.validate(ID_CACHE_SIZE, idCacheSize);
```
I wonder if this should use `Validators.GE_ZERO` so that users can disable the cache for specific addresses even if a "parent" address-setting uses a value > 0.

fartzy replied:
Yeah makes sense! changed to `Validators.GE_ZERO`

jbertram commented:
Great work here, @fartzy. Thanks for the contribution!
